### PR TITLE
PLT-1566 Add cost semantics for basic builtins

### PR DESCRIPTION
**Pull request by mjaskelioff**
Cost semantics in the metatheory for the following builtin functions, and their corresponding costing models.

```
addInteger
subtractInteger
multiplyInteger
divideInteger
quotientInteger
remainderInteger
modInteger
equalsInteger
lessThanInteger
lessThanEqualsInteger
ifThenElse
```

### CODE REVIEW

File: plutus-metatheory/src/Untyped/CEKWithCost.lagda.md [hunk 1]
```diff
@@ -34,30 +36,56 @@ open import Untyped.CEK
...
-spendBuiltin b = tell (cekMachineCost (BBuiltinApp b))
+```
+spend : StepKind → CekM ⊤
+spend st = tell (cekMachineCost (BStep st))
 
 spendStartupCost : CekM ⊤
 spendStartupCost = tell startupCost
 ```
 
+In order to calculate the cost of executing a builting, we obtain a vector with
```
Comment (michaelpj) [inline plutus-metatheory/src/Untyped/CEKWithCost.lagda.md:63]:
```suggestion
In order to calculate the cost of executing a builtin, we obtain a vector with 
```

**Review by michaelpj — APPROVED** [review]
LGTM!

File: plutus-metatheory/src/Cost.lagda.md [hunk 2]
```diff
@@ -79,93 +93,229 @@ isMonoidExBudget = record {
...
+TODO: Finish all models.
 
 ``` 
-stepKindList : List StepKind
-unquoteDef stepKindList = defListConstructors (quote StepKind) stepKindList 
+data CostingModel : ℕ → Set where 
+   -- Any number of variables 
+  constantCost       :  ∀{n} → CostingNat → CostingModel n
+  linearCostIn       : ∀{n} → Fin n → Intercept → Slope → CostingModel n
+  -- at least two arguments
```
[the remark this answers is not in the document]
Comment (mjaskelioff) [reply plutus-metatheory/src/Cost.lagda.md:191]: This a generalisation. As for higher number of arguments, constant and linear-on-a-single variable eem to be the used models. In this way I only need to provide semantics for these two models.

File: plutus-metatheory/src/Cost.lagda.md [hunk 3]
```diff
@@ -79,93 +93,229 @@ isMonoidExBudget = record {
...
+A model of a builtin consists of a pair of costing models, one for CPU and one for memory.
 
 ```
-data ExBudgetCategory : Set where
-    BStep       : StepKind → ExBudgetCategory
-    BBuiltinApp : Builtin → ExBudgetCategory  -- Cost of evaluating a fully applied builtin function
-    BStartup    : ExBudgetCategory
+record BuiltinModel (b : Builtin) : Set where 
+    field 
+        costingCPU costingMem : CostingModel (arity b)
```
[the remark this answers is not in the document]
Comment (mjaskelioff) [reply plutus-metatheory/src/Cost.lagda.md:210]: We don't have type arguments per se, so this is the number of value arguments. In the formalisation, all signatures have the type quantifiers at the beginning, so we only record them with a number (actually two numbers, so we know how many * and how many # variables there are). See the [definition of signature](https://github.com/input-output-hk/plutus/blob/e49730a5d4c1e98490985daaeab3e4386fa04e04/plutus-metatheory/src/Builtin/Signature.lagda.md?plain=1#L118C8-L118C8).

File: plutus-metatheory/src/Cost/Base.lagda.md [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+open import Data.Vec using (Vec)
+open import Relation.Binary.PropositionalEquality using (_≡_)
+
+open import Utils.Reflection using (defDec;defShow;defListConstructors)
+open import RawU using (TmCon)
+open import Builtin using (Builtin;arity)
+```
+
+We will represent costs with Naturals. In the implementation `SatInt` is used, (integers that don't overflow, but saturate). 
+As long as the budget is less than the maxInt then the result should be the same.
```
[the remark this answers is not in the document]
Comment (mjaskelioff) [reply plutus-metatheory/src/Cost/Base.lagda.md:24]: Yes. That is something we can do to be more certain about this. However, I don't think it matters for the purposes of analysing different cost models.

**Pull request merged**